src/prefect/engine.py

```diff
@@ -1267,16 +1267,40 @@ def get_state_for_result(obj: Any) -> Optional[State]:
         return flow_run_context.task_run_results.get(id(obj))
 
 
+def _link_if_trackable(obj: Any, state: State, ctx: FlowRunContext) -> None:
+    """Track connection between a task run result and its associated state if it has a unique ID.
+
+    We cannot track booleans, Ellipsis, None, NotImplemented, or the integers from -5 to 256
+    because they are singletons.
+    """
+    if (type(obj) in UNTRACKABLE_TYPES) or (
+        isinstance(obj, int) and (-5 <= obj <= 256)
+    ):
+        return
+    ctx.task_run_results[id(obj)] = state
```

Comment:
Curious why this is separated out? It makes it so you need to use `partial` and have it consume multiple additional arguments that would otherwise be available if it was written as a closure. Our current standard for defining visit functions is to do so in a closure where it's being used.

Reply:
Jeremiah suggested doing it a few comments above. I'm flexible either way.

Reply:
He noted it's not making use of variables in the closure so presumably you were using `partial` then as well? It's making use of the flow run context and state variables from the parent and it's simpler to just not use `partial` and keep the function inline. Sorry for the whiplash but I'd go back since there's not a use-case for this function as a standalone.

Reply:
I agree here -- earlier I don't think we were using `visit_collection` at all (so no partial) but now it makes sense to redefine rather than use partial just to mimic the same.

Reply:
Yeah, when I originally made the function I figured I would pass it the parameters explicitly just for the sake of clarity, and in case we ever did need to move it somewhere else. I figured the partial gave me the portability to move it elsewhere while the closure kept the pattern. But if it's never going to be used anywhere else, I can toss it back inside and just use the closure without the parameters.

Reply:
Yeah I'd avoid optimizing for the case where we might move it somewhere else before we have a need to. Following the pattern we use for `visit_fn` elsewhere should be the priority unless there's additional justification.